tools/RepositoryManagement/DevHome.RepositoryManagement/Services/RepositoryManagementDataAccessService.cs

```diff
@@ -0,0 +1,104 @@
+{
+    private readonly ILogger _log = Log.ForContext("SourceContext", nameof(RepositoryManagementDataAccessService));
+
+    private readonly IHost _host;
+
+    public RepositoryManagementDataAccessService(IHost host)
+    {
+        // Store the host to make dbContext and RepositoryManagementItemViewModel
+        // dbContext can not be passed in because RepositoryManagementDataAccessService is singleton
+        // and dbContext needs to be scoped.
+        // RepositoryManagementItemViewModel can not be passed in because multiple are made.
+        // The best solution is to make factories for DevHomeContext and RepositoryManagementItemViewModel.
+        // Might be in a future change.
+        _host = host;
+    }
+
+    public void AddRepository(string repositoryName, string cloneLocation)
+    {
+        Repository newRepo = new();
+        newRepo.RepositoryName = repositoryName;
+        newRepo.RepositoryClonePath = cloneLocation;
+
+        RepositoryMetadata newMetadata = new();
+        newMetadata.Repository = newRepo;
+        newMetadata.RepositoryId = newRepo.RepositoryId;
+        newMetadata.IsHiddenFromPage = false;
+
+        newRepo.RepositoryMetadata = newMetadata;
+
+        var dbContext = _host.GetService<DevHomeDatabaseContext>();
+        dbContext.Add(newRepo);
+        dbContext.Add(newMetadata);
+        dbContext.SaveChanges();
+    }
+
+    public List<RepositoryManagementItemViewModel> GetRepositories(bool removeRepositoriesNotInTheirSavedLocation)
+    {
+        var repos = QueryDatabaseForRepositories(removeRepositoriesNotInTheirSavedLocation);
+        return ConvertToLineItems(repos);
+    }
+
+    private IIncludableQueryable<Repository, RepositoryMetadata> QueryDatabaseForRepositories(bool removeRepositoriesNotInTheirSavedLocation)
+    {
+        var dbContext = _host.GetService<DevHomeDatabaseContext>();
+        var repositoriesFromDatabase = dbContext.Repositories.Include(x => x.RepositoryMetadata);
+
+        if (!removeRepositoriesNotInTheirSavedLocation)
+        {
+            return repositoriesFromDatabase;
+        }
+
+        // The database can get out of sync.  Take this as an example.
+        // DevHome writes a repository.
+        // The user moves or deletes the repository.
+        // Now the database record has the correct name, but the incorrect location.
+        // Remove records where the directory does not exist.
+        repositoriesFromDatabase
+            .ToList() // EF does not understand Directory.Exists.  Fetch the data first.
+            .Where(x => !Directory.Exists(x.RepositoryClonePath))
+            .ToList()
+            .ForEach(x => dbContext.Repositories.Remove(x));
+
+        dbContext.SaveChanges();
+
+        return dbContext.Repositories.Include(x => x.RepositoryMetadata);
+    }
+
+    private List<RepositoryManagementItemViewModel> ConvertToLineItems(IIncludableQueryable<Repository, RepositoryMetadata> repositories)
```

Comment:
having the database layer know about stuff (RepositoryManagementItemViewModel) "above" it seems bad.  the db shouldn't know or care about the presentation of the data imo.

this means you can't include this code in another service without pulling in ViewModels that have nothing to do with the data.

Reply:
perhaps I retract this - the ViewModel is just a strange way of saying "object with MVVM triggers".  i still maintain the stance, but maybe this is how MVVM entangles itself into a design (.. :S)